futures-util/src/task/atomic_waker.rs

```diff
@@ -198,80 +238,91 @@ impl AtomicWaker {
-                            // wake the task and return.
-                            waker.wake();
-                        }
+        let state = self.state.load(Acquire);
+        if (state == EMPTY || state == HOLDS)
+            && self.state.compare_and_swap(state, REGISTERING, Acquire) == state
+        {
+            return unsafe {
+                debug_assert_eq!(state == HOLDS, (*self.waker.get()).is_some());
+
+                // Locked acquired, update the waker cell
+                *self.waker.get() = Some(lw.clone().into_waker());
+
+                // Release the lock. If the state transitioned to include
+                // the `WAKING` bit, this means that a wake has been
+                // called concurrently, so we have to remove the waker and
+                // wake it.`
+                //
+                // Start by assuming that the state is `REGISTERING` as this
+                // is what we jut set it to.
+                let res = self.state.compare_exchange(
+                    REGISTERING, HOLDS, AcqRel, Acquire);
+
+                match res {
+                    Ok(_) => {}
+                    Err(actual) => {
+                        // This branch can only be reached if a
+                        // concurrent thread called `wake`. In this
+                        // case, `actual` **must** be `REGISTERING |
+                        // `WAKING`.
+                        debug_assert_eq!(actual, REGISTERING | WAKING);
+
+                        // Take the waker to wake once the atomic operation has
+                        // completed.
+                        let waker = (*self.waker.get()).take().unwrap();
+
+                        // Just swap, because no one could change state while state == `REGISTERING` | `WAKING`.
+                        self.state.swap(EMPTY, AcqRel);
+
+                        // The atomic swap was complete, now
+                        // wake the task and return.
+                        waker.wake();
                     }
                 }
             }
-            WAKING => {
-                // Currently in the process of waking the task, i.e.,
-                // `wake` is currently being called on the old task handle.
-                // So, we call wake on the new waker
-                lw.wake();
-            }
-            state => {
-                // In this case, a concurrent thread is holding the
-                // "registering" lock. This probably indicates a bug in the
-                // caller's code as racing to call `register` doesn't make much
-                // sense.
-                //
-                // We just want to maintain memory safety. It is ok to drop the
-                // call to `register`.
-                debug_assert!(
-                    state == REGISTERING ||
-                    state == REGISTERING | WAKING);
-            }
+        }
+
+        if state == WAKING || state == WAKING | HOLDS {
```

Comment:
This state is read at the beginning, maybe  it can't used to infer whether or not to execute lw.wake()?

Reply:
Do you mean `state` variable need to be updated with the result of `compare_and_swap`?

I think that doesn't change behaviour. However that will probably be safer and easier to read, and no performance difference. I'll update the PR.